src/audio/aec/aec.c

```diff
@@ -0,0 +1,543 @@
+	case SOF_CTRL_CMD_BINARY:
+		comp_info(dev, "aec_cmd_get_data(), SOF_CTRL_CMD_BINARY");
+		break;
+	default:
+		comp_err(dev, "aec_cmd_set_data() error: invalid command %d", cdata->cmd);
+		ret = -EINVAL;
+		break;
+	}
+	return ret;
+}
+
+static int aec_cmd_set_data(struct comp_dev *dev,
+			    struct sof_ipc_ctrl_data *cdata)
+{
+	int ret = 0;
+
+	switch (cdata->cmd) {
+	case SOF_CTRL_CMD_BINARY:
+		comp_info(dev, "aec_cmd_set_data(), SOF_CTRL_CMD_BINARY");
+		break;
+	default:
+		comp_err(dev, "aec_cmd_set_data() error: invalid command %d", cdata->cmd);
+		ret = -EINVAL;
+		break;
+	}
+	return ret;
+}
+
+/* used to pass standard and bespoke commands (with data) to component */
+static int aec_cmd(struct comp_dev *dev, int cmd, void *data,
+		   int max_data_size)
+{
+	struct sof_ipc_ctrl_data *cdata = data;
+	int ret = 0;
+
+	comp_info(dev, "aec_cmd()");
+
+	switch (cmd) {
+	case COMP_CMD_SET_DATA:
+		ret = aec_cmd_set_data(dev, cdata);
+		break;
+	case COMP_CMD_GET_DATA:
+		ret = aec_cmd_get_data(dev, cdata, max_data_size);
+		break;
+	}
+
+	return ret;
+}
+
+static int aec_trigger(struct comp_dev *dev, int cmd)
+{
+	comp_info(dev, "aec_trigger()");
+
+	return comp_set_state(dev, cmd);
+}
+
+/* copy and process stream data from source to sink buffers */
+static int aec_copy(struct comp_dev *dev)
+{
+	struct list_item *blist;
+	struct comp_buffer *sink;
+	struct comp_buffer *source;
+	int frames;
+	int num_blocks;
+	int ref_bytes;
+	int sink_bytes;
+	int source_bytes;
+	const struct audio_stream *sources_stream[AEC_MAX_SOURCES] = { NULL };
+	struct comp_data *cd = comp_get_drvdata(dev);
+	struct comp_buffer *sources[AEC_MAX_SOURCES] = { NULL };
+	uint32_t flags = 0;
+	int num_sources = 0;
+
+	comp_dbg(dev, "aec_copy()");
+
+	/* align source streams with their respective configurations */
+	list_for_item(blist, &dev->bsource_list) {
+		source = container_of(blist, struct comp_buffer, sink_list);
+		buffer_lock(source, &flags);
+		if (source->source->state == dev->state) {
+			if (num_sources < AEC_MAX_SOURCES) {
+				sources[num_sources] = source;
+				sources_stream[num_sources] = &source->stream;
+				num_sources++;
+			} else {
+				buffer_unlock(source, flags);
+			}
+		} else {
+			buffer_unlock(source, flags);
+		}
+	}
+
+	/* check if there are any sources active */
+	if (num_sources == 0 || num_sources > 2)
+		return -EINVAL;
+
+	sink = list_first_item(&dev->bsink_list, struct comp_buffer, source_list);
+	buffer_lock(sink, &flags);
+	frames = audio_stream_avail_frames(sources_stream[0], &sink->stream);
+	buffer_unlock(sources[0], flags);
+	if (num_sources == 2) {
+		frames = MIN(frames, audio_stream_avail_frames(sources_stream[1], &sink->stream));
+		buffer_unlock(sources[1], flags);
+		cd->ref_format = sources_stream[1]->frame_fmt;
+		cd->ref_active = true;
+
+		/* Needed reference shift for 32 bit filter calculate */
+		switch (cd->ref_format) {
+		case SOF_IPC_FRAME_S32_LE:
+			cd->ref_shift = 0;
+			cd->ref_32bits = true;
```

Comment:
why not make refshift negative to mark invalid, then you don't need this bool. or use a lookup table to shift bits.